flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java

```diff
@@ -405,6 +411,8 @@ public void merge(W mergeResult,
 		// windowAssigner is event time and current timestamp + allowed lateness no less than element timestamp
 		if (isSkippedElement && lateDataOutputTag != null && isElementLate(element)) {
 			sideOutput(element);
+		} else if (isSkippedElement) {
```

Comment:
I think we also need to check whether it's late. An element could also be skipped because no windows were assigned to it.

Reply:
i think when the `isSkippedElement` is true,  the `isElementLate(element)` is always be true. Because `isSkippedElement` is true when all the assigned window's window.endtime + allowLateness < currentLowWatermark, and `isElementLate` is true when element.time + allowLateness < currentLowWatermark. and element.time is <= bigest window.endtime.  so does `isElementLate` always be true when isSkippedElement is true?  And i think if i want to rule  out the situation that **because no windows were assigned to it.**, i just need to judge  whether the variable `Collection elementWindows` is empty?

Reply:
Or we can just check if `isElementLate(element)`, that's what it is there for. 😉

Reply:
haha, i know your mean now, i think its interesting ^ _ ^.  i push the code again , and modify the metrics name according to @zentol , please review the code again @aljoscha